runtime/android/core/src/org/xwalk/core/XWalkInitializer.java

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2015 Intel Corporation. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package org.xwalk.core;
+
+import android.content.Context;
+import android.os.AsyncTask;
+
+public class XWalkInitializer {
+    public abstract class XWalkInitListener {
+        public void onXWalkInitStarted() {}
+        public void onXWalkInitProgressed(int progressInPercent) {}
+        public abstract void onXWalkInitFinished();
+        public abstract void onXWalkInitFailed(XWalkInitResult result);
+    }
+
+    public enum XWalkInitResult {
+        SUCCESS,
+        ARCHITECTURE_MISMATCH,
+    }
+
+    public static void initAsync(Context context, XWalkInitListener listener) {
+        new InitializationTask(context, listener).execute();
+    }
+
+    private static class InitializationTask
+            extends AsyncTask<Void, Integer, XWalkInitResult> {
+        Context mContext;
+        XWalkInitListener mListener;
+
+        InitializationTask(Context context, XWalkInitListener listener) {
+            mContext = context;
+            mListener = listener;
+        }
+
+        @Override
+        protected void onPreExecute() {
+            mListener.onXWalkInitStarted();
+        }
+
+        @Override
+        protected XWalkInitResult doInBackground(Void... params) {
+            try {
+                XWalkCoreWrapper.initEmbeddedMode();
+            } catch (VerifyError e) {
+                return XWalkInitResult.ARCHITECTURE_MISMATCH;
+            }
+            publishProgress(50);
+
```

Comment:
Seems the decompressing work will inject here, since now we don't know how long it will consume, I'm not sure whether this progress update is meaningful.